cluster-autoscaler/proposals/plugable-provider-grpc.md

```diff
@@ -0,0 +1,608 @@
+
+  rpc PricingNodePrice(PricingNodePriceRequest)
+    returns (PricingNodePriceResponse) {}
+
+  rpc PricingPodPrice(PricingPodPriceRequest)
+    returns (PricingPodPriceResponse)
+
+  rpc NewNodeGroup(NewNodeGroupRequest)
+    returns (NewNodeGroupResponse) {}
+
+  rpc GetResourceLimiter(GetResourceLimiterRequest)
+    returns (GetResourceLimiterResponse) {}
+
+  rpc GPULabel(GPULabelRequest)
+    returns (GPULabelResponse) {}  
+
+  rpc GetAvailableGPUTypes(GetAvailableGPUTypesRequest)
+    returns (GetAvailableGPUTypesResponse) {}   
+
+  rpc Cleanup(CleanupRequest)
+    returns (CleanupResponse) {}        
+
+  rpc Refresh(RefreshRequest)
+    returns (RefreshResponse) {}
+
+  // NodeGroup specific RPC functions
+  rpc NodeGroupTargetSize(NodeGroupTargetSizeRequest)
+    returns (NodeGroupTargetSizeResponse) {}        
+
+  rpc NodeGroupIncreaseSize(NodeGroupIncreaseSizeRequest)
+    returns (NodeGroupIncreaseSizeResponse) {}    
+
+  rpc NodeGroupDeleteNodes(NodeGroupDeleteNodesRequest)
+    returns (NodeGroupDeleteNodesResponse) {}    
+
+  rpc NodeGroupDecreaseTargetSize(NodeGroupDecreaseTargetSizeRequest)
+    returns (NodeGroupDecreaseTargetSizeResponse) {}         
+
+  rpc NodeGroupNodes(NodeGroupNodesRequest)
+    returns (NodeGroupNodesResponse) {}              
+
+  rpc NodeGroupDelete(NodeGroupDeleteRequest)
+    returns (NodeGroupDeleteResponse) {}          
+
+  rpc NodeGroupCreate(NodeGroupCreateRequest)
+    returns (NodeGroupCreateResponse) {}
+
+  rpc NodeGroupTemplateNodeInfo(NodeGroupDTemplateNodeInfoRequest)
+    returns (NodeGroupTemplateNodeInfoResponse) {}                                                               
+}
+```
+
+Note that, the rest of message used in these calls are detailed in the [Appendix](#appendix) section.
+
+In order to talk to the custom cloud provider server, this new cloud provider has to be registered
+when bootstrapping the CA.
+Consequently, the CA needs to expose new flags to specify the cloud provider and all the required properties
+to reach the remote gRPC server.
+
+A new flag, named
+`--cloud-provider-url=https://local.svc.io/mycloudprovider/server`, determines the URL to reach the custom provider implementation.
+In addition this approach reuses the existing flag that defines the name of the cloud provider using a pre-defined value `--cloud-provider=external`
+https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler/cloudprovider. `external` defines the usage
+of an external cloud provider whose interface is handled by a remote gRPC service.
+
+To connect the CA core with this new external cloud provider, this approach needs to implement a new generic cloud provider
+as part of the CA core code.
+This new provider, named `ExternalCloudProvider`, makes gRPC calls to the remote functions exposed by the external cloud provider server. In other words, it forwards the calls and handle the errors analogously how done in other existing providers.
```

Comment:
similar to previous comment, we should probably name this `ExternalGrpcCloudProvider`

Reply:
yes, i'll change it accordingly.